### PR TITLE
BlueFS: fix unlogged updating of fnode and race condition

### PR DESCRIPTION
**Pull request by xiexingguo**
And some fixes for the sake of performance.

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas Gathered https://github.com/ceph/ceph/pull/7308 here and add some more fixes.

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -741,8 +742,6 @@ int BlueFS::_read(
     }
 
     dout(30) << __func__ << " result chunk (" << r << " bytes):\n";
-    bufferlist t;
-    t.substr_of(buf->bl, off - buf->bl_off, r);
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: Your patch might have substring taken from the different offset comparing to the original code

**Comment by xiexingguo** [comment]
@liewegas  And this PR too:-)

**Comment by xiexingguo** [comment]
@liewegas All fixed.

**Comment by liewegas** [comment]
Looks good, thanks!

**Comment by liewegas** [comment]
this conflicts with #7383 , can you rebase on top of that?  thanks!

**Comment by liewegas** [comment]
(Mutex is replaced by std::mutex)

**Comment by xiexingguo** [comment]
OK, I will rebase as soon as it is merged.

**Comment by xiexingguo** [comment]
Rebased.

**Comment by xiexingguo** [comment]
@liewegas  Rebased.

**Comment by liewegas** [comment]
merged manually

**Pull request closed without merging**